log4j-api/src/main/java/org/apache/logging/log4j/message/ReusableParameterizedMessage.java

```diff
@@ -355,4 +356,41 @@ public void clear() { // LOG4J2-1583
         messagePattern = null;
         throwable = null;
     }
+
+    @Override
+    public MessageContentFormatter getMessageContentFormatter() {
+        return Formatter.INSTANCE;
+    }
+
+    private enum Formatter implements MessageContentFormatter {
+        INSTANCE;
+
+        private static final ThreadLocal<int[]> localIndices = new ThreadLocal<int[]>() {
+            @Override
```

Comment:
This pattern (subclassing ThreadLocal to provide an initial value) is problematic for frameworks that use class loaders. If  I remember correctly, Tomcat reports this as a potential memory leak. The reason is that the ThreadLocal map now has a reference to a Log4j class (the anonymous ThreadLocal subclass) that is never removed, preventing the garbage collector from GC-ing a web app that was stopped or restarted.

A better way is to simply declare the threadlocal as

```java
private static final ThreadLocal<int[]> localIndices = new ThreadLocal<int[]>();
```
... and providing an accessor (getter) method where the initial value is set. The code using this threadlocal must always use the accessor. This is also not great, but prevents the memory leak.

Reply:
Thanks. @bjlaub if you are still interested is it possible you could make the recommended changes?